cpp/src/arrow/csv/writer.cc

```diff
@@ -128,10 +139,18 @@ class ColumnPopulator {
     // Populators are intented to be applied to reasonably small data.  In most cases
     // threading overhead would not be justified.
     ctx.set_use_threads(false);
-    ASSIGN_OR_RAISE(
-        std::shared_ptr<Array> casted,
-        compute::Cast(data, /*to_type=*/utf8(), compute::CastOptions(), &ctx));
-    casted_array_ = checked_pointer_cast<StringArray>(casted);
+    if (data.type() && is_large_binary_like(data.type()->id())) {
+      ASSIGN_OR_RAISE(array_, compute::Cast(data, /*to_type=*/large_utf8(),
+                                            compute::CastOptions(), &ctx));
+    } else {
+      auto casted = compute::Cast(data, /*to_type=*/utf8(), compute::CastOptions(), &ctx);
```

Comment:
wonder which scenerio would this happens? When a extremly huge array is passed in `data`? Should an error code being checked here or just not casting it?

Reply:
If we raise, user would have no way to _easily_ write their tables to CSV without hacky split operations.

Reply:
I mean, can we only check `Status::CapacityError` here?

Reply:
I see. Fixed now.